src/SFA.DAS.EAS.Portal/Application/Commands/Reservation/AddReservationCommand.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.EAS.Portal.Application.Services;
+using SFA.DAS.EAS.Portal.Database.Models;
+using SFA.DAS.Reservations.Messages;
+
+namespace SFA.DAS.EAS.Portal.Application.Commands.Reservation
+{
+    public class AddReservationCommand
+    {
+        private readonly IAccountDocumentService _accountsService;
+        private readonly ILogger<AddReservationCommand> _logger;
+
+        public AddReservationCommand(IAccountDocumentService accountsService, ILogger<AddReservationCommand> logger)
+        {
+            _accountsService = accountsService;
+            _logger = logger;
+        }
+
+        public async Task Execute(ReservationCreatedEvent reservedFunding, string messageId, CancellationToken cancellationToken = default)
+        {
+            // TODO: move logging to a decorator
+            _logger.LogInformation("Executing AddReservationCommand");
+
+            var accountDocument = await _accountsService.Get(reservedFunding.AccountId);
+
+            if (accountDocument == null)
+            {
+                accountDocument = AccountDocument.Create(reservedFunding.AccountId);
+                accountDocument.Account.Name = reservedFunding.AccountLegalEntityName;
+
+                var newOrg = new Types.Organisation() {
+                    Id = reservedFunding.AccountLegalEntityId,
+                    Name = reservedFunding.AccountLegalEntityName };
+
+                accountDocument.Account.Organisations.Add(newOrg);
+
+                newOrg.Reservations.Add(new Types.Reservation()
+                {
+                    Id = reservedFunding.Id,
+                    CourseCode = reservedFunding.CourseId,
+                    CourseName = reservedFunding.CourseName,
+                    StartDate = reservedFunding.StartDate,
+                    EndDate = reservedFunding.EndDate
+                });
+            }
+            else
+            {
+                var org = accountDocument.Account.Organisations.Where(o => o.Id.Equals(reservedFunding.AccountLegalEntityId)).First();
+
+                var existing = org.Reservations.FirstOrDefault(r => r.Id.Equals(reservedFunding.Id));
+                if (existing != null)
+                {
+                    return;  // already handled 
```

Comment:
We should log a warning if we ever handle a message and we already have a reservation with its id. In theory, this should _never_ happen, as we de-dupe messages and hopefully reservations won't be sending us messages with different ids and identical reservation ids.

Reply:
Do we care?

Reply:
If the system is behaving against expectations, then either our understanding of the system is wrong, or there is a bug in the system, either way, worth knowing.

Reply:
But the service bus would log this surely.  This is handling a failure of the messaging not the application.

Reply:
If we were throwing an exception, then yes that would be logged, but it seems we are silently ignoring the issue, so we'd have no clue that it'd happened. (It may be possible to piece it together from what we do log, but it wouldn't be easy to notice that it'd happened.)

Reply:
I agree with logging a warning, another reason being this may not be simply a "failure of the messaging". We don't know how the reservations side may change, it may be in error but they may publish multiple events with the same reservation and it would be handy to be able to trace such a thing via the logs

Reply:
But it's the AddReservation command.  Why would the reservation service publish that event twice?  We ignore it because we may receive it twice if the messaging system sends to the handler again because we took too long to process it.

Reply:
As @cofaulco mentioned, the reservations service might have an error (due to a bug).

We don't have to explicitly handle the situation you described. We handle idempotency in a cross-cutting manner. We have built in message de-duplication, through the use of the outbox pattern, for messages processed serially, and etag handling for messages processed concurrently (although the etag handling is more of an 'eventual idempotency', as the concurrent write will fail, and the retry will be quietly consumed due to the messageId already being processed).

It probably would be better to throw an exception, rather than log a warning, as it's less likely to be missed.